server/vulnerabilities/oval/analyzer.go

```diff
@@ -17,73 +17,124 @@ import (
 	"github.com/hashicorp/go-multierror"
 )
 
+const (
+	hostsBatchSize = 500
+	vulnBatchSize  = 500
+)
+
 // Analyze scans all hosts for vulnerabilities based on the OVAL definitions for their platform,
 // inserting any new vulnerabilities and deleting anything patched.
 func Analyze(
 	ctx context.Context,
 	ds fleet.Datastore,
-	versions *fleet.OSVersions,
+	ver fleet.OSVersion,
 	vulnPath string,
 ) ([]fleet.SoftwareVulnerability, error) {
+	platform := NewPlatform(ver.Platform, ver.Name)
+
+	if !platform.IsSupported() {
+		return nil, nil
+	}
+
+	defs, err := loadDef(platform, vulnPath)
+	if err != nil {
+		return nil, err
+	}
+
 	var errR error
 
-	for _, v := range versions.OSVersions {
-		platform := NewPlatform(v.Platform, v.Name)
-		if !platform.IsSupported() {
-			continue
-		}
+	// Since hosts and software have a M:N relationship, the following sets are used to
+	// avoid doing duplicated inserts/delete operations (a vulnerable software might be
+	// present in many hosts).
+	toInsertSet := make(map[string]fleet.SoftwareVulnerability)
+	toDeleteSet := make(map[string]fleet.SoftwareVulnerability)
+
+	var offset int
+	for {
+		hIds, err := ds.HostIDsByOSVersion(ctx, ver, offset, hostsBatchSize)
+		offset += hostsBatchSize
 
-		defs, err := loadDef(platform, vulnPath)
 		if err != nil {
 			errR = multierror.Append(errR, err)
 			continue
```

Comment:
You could end up retrying hard in case of db errors, maybe just return the error?

Reply:
Sounds good

Reply:
Done